tests/suite/cli-ui/rustup/rustup_only_options_stdout.toml

```diff
@@ -1,48 +1,13 @@
 bin.name = "rustup"
 args = ["-v"]
 status.code = 1
-stdout = """
-rustup [..]
-The Rust toolchain installer
+stdout = ""
+stderr = """
+error: error: 'rustup' requires a subcommand but one was not provided
```

Comment:
This change is surprising. `rustup -v` should print the version and basic help, not an error about subcommand selection.

Reply:
https://github.com/clap-rs/clap/discussions/5076

Reply:
I agree with @epage's point in https://github.com/clap-rs/clap/discussions/5076#discussioncomment-6752420 and I don't even know what `rustup -v` is supposed to do personally... 👀

Reply:
If this test had instead used `-y` or `--no-modify-path`, would the answer be any different of what this should do?

Reply:
This is a new test added during the clap-3 migration (specifically if I remember correctly during the upgrade aftermath when we got folk to test and report on bugs).

This test was added in https://github.com/rust-lang/rustup/pull/3291 referencing https://github.com/rust-lang/rustup/issues/3290.

I think it is reasonable to expect that `rustup` on its own with only globally scoped, debug-affecting flags like `-v` or `-q` should behave the same as only `rustup` - or perhaps output something more or less such as startup debug messages.

Reply:
@rbtcollins The point here is, as the code is currently written, `rustup` in isolation has no clear meaning. Could we make a default (but hidden) subcommand to just print the help string?

Reply:
@rami3l I don't entirely follow?

`rustup` as shipped today - not this PR, but the in use binary, has the following behaviour:
`$ rustup` -> outputs help
`$ rustup -v` -> outputs help ('verbosely' - no actual change)
`$ rustup -q` -> outputs help ('verbosely' - no actual change)

This PR changes that to an error message that is somewhat inscrutable and certainly doesn't guide users to realise that they need to remove the `-v`/`-q`. 

So what options do we have? We could:

- stay on clap 3.
  -  Pros: UI stays the same.
  -  Cons: its dead upstream as 4 is released and we still have that attr CVE.
- accept the current behaviour in this PR. 
  - Pros: we get future bug fixes from the clap developers
  - Cons: we have to deal with bug reports from confused users
- find a way to retain the existing behaviour with clap 4.
  - Pros: we get both the pros of the first two options
  - Cons: more development is needed and its not clear how to do it

Reply:
@rbtcollins I meant we should implement that help message-printing behavior explicitly as `rustup`'s default behavior, rather than relying on `arg_required_else_help` whose semantics has already slightly shifted. I'm not sure if this can be done properly though, but I'll find some time to see if I can get a demo working.

Reply:
@djc This patch does seem to work:

```diff
diff --git a/src/cli/rustup_mode.rs b/src/cli/rustup_mode.rs
index 001f86b7..644ecd6b 100644
--- a/src/cli/rustup_mode.rs
+++ b/src/cli/rustup_mode.rs
@@ -243,7 +243,10 @@ pub fn main() -> Result<utils::ExitCode> {
             }
             _ => unreachable!(),
         },
-        None => unreachable!(),
+        None => {
+            cli().print_long_help()?;
+            utils::ExitCode(0)
+        }
     })
 }
 
@@ -253,8 +256,7 @@ pub(crate) fn cli() -> Command {
         .about("The Rust toolchain installer")
         .before_help(format!("rustup {}", common::version()))
         .after_help(RUSTUP_HELP)
-        .subcommand_required(true)
-        .arg_required_else_help(true)
+        .subcommand_required(false)
         .arg(
             verbose_arg("Enable verbose output"),
         )
```

Reply:
@rami3l thanks, looks great! I changed the `ExitCode` to 1 since I think we should still technically signal this as an error, and pushed the branch again to see if it can pass CI this time.

Reply:
> If this test had instead used `-y` or `--no-modify-path`, would the answer be any different of what this should do?

IMO yes, in that case it should also act according to the "missing subcommand" logic.

Reply:
@djc That's interesting. I can reproduce this (see the output above) but I have no idea what this is about, sorry :/

Reply:
Just generic output from `-v` at initialization, I think. CI fails in yet another way:

https://github.com/rust-lang/rustup/actions/runs/5927865297/job/16072261180?pr=3444#step:17:632

Reply:
> `rustup` as shipped today - not this PR, but the in use binary, has the following behaviour: `$ rustup` -> outputs help `$ rustup -v` -> outputs help ('verbosely' - no actual change) `$ rustup -q` -> outputs help ('quietly' - no actual change)
> 
> This PR changes that to an error message that is somewhat inscrutable and certainly doesn't guide users to realise that they need to remove the `-v`/`-q`.

imo this analysis is backwards.  I do not see why a user would pass in `rustup -v` and expect help output.  To me, this feels like we are overall making a mountain out of a molehill and putting in more thought and code design work to something that 99% won't even notice or care about.

Reply:
The metadata output is the result of -v having effect; I suggest changing to -q for the test, which is effectively guaranteed to never be adding more output.

@epage I can appreciate the sentiment about engineering effort / reward. But it seems to me that you take issue with the value judgement not the analysis - you haven't disputed the historical behaviour, argued that the error message is easily understood or proposed alternative ways forward. 

So if the question is about the relative merits of option 2 vs option 3, what I can say right now is that `rustup` is under-resourced in terms of human capacity. And triaging and improving non-functional bugs takes quite a bit of time. We could of course say 'if it isn't a functional problem, we will just ignore/close such bugs'.... but I don't feel that that would be a good way of upholding the excellent work that preceeded my joining the project a few years back.

Perceptions of quality in software often hinge on non-happy-path, non-functional-requirements. We know that at least some people went and tried this path, or we wouldn't have the additional test added to cover it. With it covered, we won't get another bug report about it, which is good. 

And @djc seems to have quite rapidly figured out and solved the problem, so thats nice.

Reply:
Well, CI is failing.

Reply:
I take issue with the analysis itself but since it started at the root of the analysis, I didn't go further into it.  I didn't just remove `SubcommandRequiredElseHelp` for code size and binary size.

Reply:
And for a project that is under resourced, flowing against the stream seems like it'd be the opposite of what you'd want to be doing.  I'm not saying this in terms of accepting bugs but of practices

Reply:
I think it depends on where we think our time is going no? We're getting quite meta at this point though.